Caught and fixed: remove singularities from result of continuous_domain()

akash9712 opened the pull request:
Continuous_domain returned wrong result for certain functions, where singularities
were included in the continuous_domain.
Example(Before fix):
```
In [3]: continuous_domain(1/x - 2, x, S.Reals)
Out[3]: ℝ
```
Added condition to continuous_domain in sympy.calculus.util to remove
singularities that were being included in the domain before. Added tests to
check for the added constraints.
Example(After fix):
```
In [2]: continuous_domain(1/x - 2, x, S.Reals)
Out[2]: (-∞, 0) ∪ (0, ∞)

In [3]: continuous_domain(1/(x**2 - 1) - 2, x, S.Reals)
Out[3]: (-∞, -1) ∪ (-1, 1) ∪ (1, ∞)
```
<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

akash9712 committed:
continuous_domain: remove singularities from result

Added condition to continuous_domain in sympy.calculus.util to remove
singularities that were being included in the domain before. Added tests to
check for the added constraints.

Example:

In [2]: continuous_domain(1/x - 2, x, S.Reals)
Out[2]: (-∞, 0) ∪ (0, ∞)

In [3]: continuous_domain(1/(x**2 - 1) - 2, x, S.Reals)
Out[3]: (-∞, -1) ∪ (-1, 1) ∪ (1, ∞)

gxyd commented on sympy/calculus/util.py:
```diff
@@ -61,15 +63,18 @@ def continuous_domain(f, symbol, domain):
         if f.has(Abs):
-            sings = solveset(1/f, symbol, domain)
+            sings = solveset(1/f, symbol, domain) + \
+                solveset(denom(together(f)), symbol, S.Reals)
```
Should this be solved in `S.Reals` or `domain`?

akash9712 replied:
Yeah, it should be `domain` instead. Updated the PR.

gxyd commented on sympy/calculus/util.py:
```diff
@@ -61,15 +63,18 @@ def continuous_domain(f, symbol, domain):
+                        solveset(denom(together(f)), symbol, S.Reals)
                     break
             else:
-                sings = Intersection(solveset(1/f, symbol), domain)
```
also in general, I think we could use `solveset(1/f, symbol, domain)` instead of using explicit intersection.

akash9712 replied:
The explicit intersection seems intentional. While solving for certain functions using `solveset`, for e.g. `1/tan(x)`, the function returns an ImageSet union if the domain is `S.Complexes` or `S.Reals`, but returns an empty set in bounded domains like `Interval(0, pi)`. Solving the function in the complex plain and taking the intersection with the domain solves this problem.

akash9712 committed:
Changes to continuous_domain()

Changed `S.Reals` to `domain` as provided in the arguments while solving for
the denominator of the specified function. Also changed previously added
imports to top-level imports.

akash9712 commented on sympy/calculus/util.py:
```diff
@@ -38,13 +38,15 @@ def continuous_domain(f, symbol, domain):
     from sympy.solvers.inequalities import solve_univariate_inequality
     from sympy.solvers.solveset import solveset, _has_rational_power
+    from sympy.simplify.radsimp import denom
+    from sympy.polys.rationaltools import together
```
No, there is no problem in changing the imports. I have updated the PR.

smichr commented on sympy/calculus/util.py:
```diff
@@ -61,15 +62,18 @@ def continuous_domain(f, symbol, domain):
             else:
-                sings = Intersection(solveset(1/f, symbol), domain)
+                sings = Intersection(solveset(1/f, symbol), domain) + \
+                    solveset(denom(together(f)), symbol, domain)
```
Does this 2nd part need to be `Intersection(solveset(denom(together(f)), symbol), domain)`, too?

akash9712 replied:
The current code seems to work correctly, though I am not very sure why the intersection is not needed in the second part. The reason, I think might be that the need to take an intersection is arising only when the function is a negative power of a trignometric function. While solving for `denom(together(f))`, there is no negative power, as any negative powers in the denominator are sent to the numerator. For example:
```
>>>f = 1/(1/sin(x) + 1/cos(x))
>>> denom(together(f))
sin(x) + cos(x)
```

smichr commented on sympy/calculus/util.py:
```diff
@@ -38,13 +37,15 @@ def continuous_domain(f, symbol, domain):
     """
     from sympy.solvers.inequalities import solve_univariate_inequality
     from sympy.solvers.solveset import solveset, _has_rational_power
+    from sympy import denom
```
Is there any reason not to use the full path to these functions as in the two lines above?

akash9712 replied:
I am sorry, I misinterpreted one of the earlier reviews regarding the imports. I am now changing it so as to specify the complete paths, and moving them to top level imports, as recommended by in previous reviews.

akash9712 committed:
Modified import statement

Specified complete path for the imports added and moved them to top level
imports.

akash9712 commented:
@gxyd @smichr The test failures seem unrelated to the changes made in the last commit, prior to which the tests passed. The tests pass locally on the system, on ground type python. Will a restart on the tests help?

ylemkimon commented:
Closing and reopening the PR triggers new Travis build.

akash9712 commented:
@ylemkimon Thank you.
